### ISSUE TITLE
test-embed `kevent` call needs a non-empty event list, otherwise spins 100%

### ISSUE THREAD

**Issue by mitchellh** [libuv/libuv#3606]
* **Version**: https://github.com/libuv/libuv/commit/9e59aa1bc8c4d215ea3e05eafec7181747206f67
* **Platform**: `Darwin Mitchells-MBP.localdomain 21.4.0 Darwin Kernel Version 21.4.0: Fri Mar 18 00:46:32 PDT 2022; root:xnu-8020.101.4~15/RELEASE_ARM64_T6000 arm64`

The example for embedding a libuv event loop into another in `test-embed.c` calls `kevent` in the following way: 

https://github.com/libuv/libuv/blob/9e59aa1bc8c4d215ea3e05eafec7181747206f67/test/test-embed.c#L80

At least on my platform, this causes `kevent` to always return _immediately_, effectively making the embed thread a hot loop.

The FreeBSD `man kevent(2)` pages say the following:

> When nevents is	zero, kevent() will return immediately even if there is	a timeout specified unlike [select(2)](https://www.freebsd.org/cgi/man.cgi?query=select&sektion=2&apropos=0&manpath=FreeBSD+13.0-RELEASE+and+Ports).

Despite the macOS man pages **NOT** including that sentence, this is also the behavior I'm seeing on Mac (at least as of the platform version above). I haven't tested on any other kqueue-enabled platforms.

The solution is to specify a non-empty event list:

```c
      struct timespec ts;
      ts.tv_sec = timeout / 1000;
      ts.tv_nsec = (timeout % 1000) * 1000000;

      struct kevent ev;
      r = kevent(fd, NULL, 0, &ev, 1, &ts);
```

This causes `kevent` to block properly.

**Comment by bnoordhuis** [comment]
Your analysis looks correct to me but I observe that the timeout in the test is always zero so it's busy looping anyway (in theory; in practice it only seems to go through a single loop iteration.)

That test is pretty poor and should probably be rewritten. It's multi-thread unsafe (accesses the event loop from multiple threads) and lacks error checking.

**Comment by mitchellh** [comment]
Oh yes, I actually didn't run the tests directly. I wasn't familiar with kqueue APIs directly and I was embedding libuv based on this as an example and I noticed on Mac it didn't work as expected. 😄 So I was running non-test code based on the test as an example, there may very well be other issues with the test itself.

Actually, another issue I noticed is that if timeout is -1 -- which happens at times -- we should be passing in `null` instead of `&ts`. Not an issue for the test most likely, but in case anyone googles this and finds this, that is another change that needs to happen for real world usage.
